app/controllers/architectures_controller.rb

```diff
@@ -39,4 +56,55 @@ def destroy
       process_error
     end
   end
+
+  private
+
+  def can_edit?(id)
+    authorized_for(controller: :architectures, action: :update, id: id)
+  end
+
+  def can_delete?(id)
+    authorized_for(controller: :architectures, action: :destroy, id: id)
+  end
+
+  def row_permissions(a)
+    { canEdit: can_edit?(a), canDelete: can_delete?(a) }
+  end
+
+  def columns
+    [
+      {
+        sortKey: 'name',
+        title: _('Name'),
+        width: 15,
+      },
+      {
+        title: _('Operating systems'),
+        width: 70,
+      },
+      {
+        title: _('Hosts'),
+        width: 15,
+      },
+    ]
+  end
+
+  def rows(architectures)
+    rows = []
+    architectures.map do |a|
+      name = can_edit?(a) ? { type: 'link', label: a.name, path: edit_architecture_path(:id => a) } : a.name
+      os = a.operatingsystems.map(&:to_label).to_sentence
+      hosts_count = { type: 'link', label: hosts_count(:architecture)[a], path: hosts_path(:search => "architecture = #{a}") }
+      rows << { cells: [name, os, hosts_count], permissions: row_permissions(a) }
+    end
+    rows
```

Comment:
You don't need the `rows` variable. You can write
```ruby
def rows(architectures)
  architectures.map do |architecture|
    # ...
    { cells: [name, os, hosts_count], permissions: row_permissions(a) }
  end
end

Reply:
thanks! any advice on how plugins could extend it in the easiest way?

Reply:
nice! thanks @timogoebel

Reply:
@ShimShtein: The more I think about this, I'm asking myself if we're reinventing the graphql api.

Reply:
Hm. I need to think more about this. Isn't this actually the view layer? I get, why you're asking for templates. Shouldn't this be done in react/JSX?

Reply:
It's kind of a middle layer: this is the data that is passed to the UI, but it's not dictated by the UI (as in graphql API). 
IMHO, graphql is a bit different approach - it gives more control about which data to get from the server to the UI components. So for graphql enabled components, the whole extensibility would look a bit different. The components (that are supplied by plugins and core) on the UI side would "vote" which fields they need from the server, and on the server side, each plugin will extend the type to supply those fields.

Reply:
Yep, true. I believe graphql (or our rest api) should supply the data. We just need a layer to define what and how the data should be presented in the UI (do we need a presenter, perhaps?)

Reply:
@timogoebel since we have extensible UI, I think presenters will not be that good of a fit. Maybe something more data-driven, something along the lines of "here is a JSON node with data, use whatever component that fits the most to render it".
In Microsoft world, the components were registered per actual class, we could theoretically mimic the same with a special property (sort of `_object_type`).

Reply:
would it make sense to handle the extendability in a follow-up PR? opened [redmine](https://projects.theforeman.org/issues/32950)

@ShimShtein started to help me with that, and we wrote some pseudo-code over a discussion in https://github.com/laviro/foreman/pull/9

The experimental architecture page won't affect the existing one,
and I am planning to do it in small steps so it will be easier also to review